File: companies/Chime/210.course-schedule-ii.py
"""
Approach: Kahn's topological sort and return the produced order.
Data structure: adjacency lists plus indegree counts model the prerequisite graph.
Interview logic: a course can be output only when all prerequisites have been output. If a cycle exists, some courses never reach indegree zero, so return an empty list.
Complexity: O(V + E) time and space.
Tests and edge cases: multiple valid orders are acceptable; no prerequisites can return 0..n-1; cycles return [].
"""
from __future__ import annotations
from collections import deque
from typing import List

# @lc code=start
from collections import deque
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        graph = [[] for _ in range(numCourses)]
        indegree = [0] * numCourses
        for course, pre in prerequisites:
            graph[pre].append(course)
            indegree[course] += 1
        q = deque(i for i, deg in enumerate(indegree) if deg == 0)
        order = []
        while q:
            node = q.popleft()
            order.append(node)
            for nxt in graph[node]:
                indegree[nxt] -= 1
                if indegree[nxt] == 0:
                    q.append(nxt)
        return order if len(order) == numCourses else []
# @lc code=end

File: Oracle/210.course-schedule-ii.cpp
#include <algorithm>
#include <array>
#include <cctype>
#include <climits>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;


class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        /*
        Approach: Kahn topological sort. Start with indegree-zero courses, append
        each removed course to the order, and reduce indegrees of courses that
        depend on it. A cycle exists if fewer than numCourses are output.

        Complexity: O(V + E) time, O(V + E) space.
        */
        vector<vector<int>> graph(numCourses);
        vector<int> indegree(numCourses, 0);
        for (auto& edge : prerequisites) {
            graph[edge[1]].push_back(edge[0]);
            ++indegree[edge[0]];
        }
        queue<int> q;
        for (int i = 0; i < numCourses; ++i) if (indegree[i] == 0) q.push(i);
        vector<int> order;
        while (!q.empty()) {
            int node = q.front(); q.pop();
            order.push_back(node);
            for (int nxt : graph[node]) if (--indegree[nxt] == 0) q.push(nxt);
        }
        return order.size() == (size_t)numCourses ? order : vector<int>{};
    }
};
